### ISSUE TITLE
Complex attribute values (json, array)

### ISSUE THREAD

**Issue by MortalFlesh** [cmeeren/Felicity#4]
Hello,

is it possible to define an attribute as a `JsonObject` or an array?

I'd like to have a resource with attributes like this:
```json
{
    ...
    "attributes": {
        "contact": {
            "email": "<email>",
            "phone": "<phone>"
        }
    }
}
```

Thank you.

**Comment by cmeeren** [comment]
Yes. In Felicity, you are free to return whatever you want. For example, if you use `define.Attribute.Simple()`, your objects will be serialized and deserialized exactly as-is.

However, please consider if this is good resource design. Why not have `email` and `phone` as direct resource attributes, so they can be updated individually? (In your example, there is no way to update just `email` or just `phone`; you have to replace the whole object, because JSON:API doesn't allow for anything else.) Alternatively, you could create a `contact` resource and have a relationship to that from the resource you are showing above.

**Comment by MortalFlesh** [comment]
Hello,

yes, I know that flatten structure is easier to handle, but I need a Domain specific types to be there.

`Contact` is a part of ubiquitous language and as a value object (so it should not be a separate resource I'd say) it cannot be changed partially. Also it should be skipped in the response, where only a resource id should be.

But I still can't handle it.
I was just able to handle a simple string, in the way you described, but it is not what I want.
```json
{
    ...
    "attributes": {
        "contact": "... contact ..."
    }
}
```

but when I send a 
```json
{
    ...
    "attributes": {
        "contact": {
            "email": "<email>",
            "phone": "<phone>"
        }
    }
}
```

I could not get a data - I always get the error 
```
{
  "errors": [
    {
      "id": "86683d8b55e54cd18a26a6520af40c92",
      "status": "400",
      "title": "Invalid request body",
      "detail": "Invalid JSON for attribute 'contact': Cannot get the value of a token type 'StartObject' as a string.",
      "source": {
        "pointer": "/data/attributes/contact"
      }
    }
  ]
}
```

**Comment by cmeeren** [comment]
Hm, not sure what's wrong, but it may look like it infers the attribute type to be string. Could you share the Felicity code for this attribute definition, and any signatures for the functions you're using in the attribute definition?

**Comment by MortalFlesh** [comment]
Well it might be the problem.

### TLDR
I was not sure what should be the type of such attribute. So I had a `string` there. I've changed it to `obj` and in the end to `System.Text.Json.JsonElement` now, and it works.

---
### Business Logic
This API is for identifying persons by contact - client send (POST) contact (json above) and gets PersonId (it would create a person if there is not one yet). So the contact should be skipped in the response.

### This is my code
_a bit simplified, but basically this ..._

```fs
// Domain

type PersonId = PersonId of string
type Email = Email of string
type Phone = Phone of string

type Contact =
    | IsEmail of Email
    | IsPhone of Phone
    | IsEmailAndPhone of Email * Phone

type Identity = {
    Id: PersonId
    Contact: Contact
}

// Resource
let contact =
        define.Attribute
            .Simple()
            .GetSkip(fun ctx identity ->
                printfn "skip: %A" identity
                Skip
            )
            .Set(fun (contact: obj) identity ->
                printfn "set: %A (%A)" c (c.GetType())
                identity
            )

let post =
        define.Operation
            .Post(fun ctx parser ->
                let require attribute =
                    attribute
                    |> parser.GetRequired
                    |> Hopac.Job.toAsync <@> AttributeErrors

                asyncResult {
                    let! (contact: obj) = require contact
                    printfn "contact in post: %A (%A)" contact (contact.GetType())

                    (* let! contact =
                        contact
                        |> Contact.parseString
                        |> AsyncResult.ofResult <@> ParseContactError *)

                    // I'd like to parse contact here, but now I just create a dummy one, to make it all pass and log only
                    let contact = IsEmail (Email "... dummy email ...")

                    return {
                        Id = PersonId "... todo ..."
                        Contact = contact
                    }
                }
                |> parser.ForAsyncRes
            )
            .AfterCreate(fun identity -> identity)
```

where `<@>` is `AsyncResult.mapError`

for
```json
{
    "data": {
        "type": "identities",
        "attributes": {
            "contact": { "email": "" }
        }
    }
}
```
the output is:
```
Post ...
contact in post: ... contact ... (System.Text.Json.JsonElement)
set: { "email": "" } (System.Text.Json.JsonElement)
skip: { Id = PersonId "...todo..."
  Contact = IsEmail Email "...email..." }
```

So when I've changed `obj` to `System.Text.Json.JsonElement`, it works and I can handle it myself.

Thanks for help!

**Comment by cmeeren** [comment]
The type used in `Simple` is what is directly serialized and deserialized. You should in general not use a DU type here, because they have no standard JSON representation. In fact you should ideally not use a domain object directly at all, because then you need to be aware that the domain object is serialized directly.

The ideal solution in your case is to create a record type for the JSON representation, e.g.

```f#
type ContactDto =
  { email: string option
    phone: string option }
```

(`option` is fine, it will be serialized as expected, as will `Skippable`, but you can also just use `string` and check for nulls)

and then map between this DTO and whatever else you have in your domain logic. That way, you have a stable JSON representation that won't accidentally change if you refactor/rename domain stuff.

You should not need to use `JsonElement`.

**Comment by MortalFlesh** [comment]
Ohh.. Now I get it :) 
The `ContactDto` type is basically what I needed from the begging.

Thank you! 👍
